arm/Microsoft.Network/azureFirewalls/deploy.bicep

```diff
@@ -102,20 +114,59 @@ param diagnosticMetricsToEnable array = [
   'AllMetrics'
 ]
 
-var ipConfigurations_var = [for ipConfiguration in ipConfigurations: {
+@description('Optional. The name of the diagnostic setting, if deployed.')
+param diagnosticSettingsName string = '${name}-diagnosticSettings'
+
+var additionalPublicIpConfigurations_var = [for ipConfiguration in additionalPublicIpConfigurations: {
   name: ipConfiguration.name
   properties: {
     publicIPAddress: contains(ipConfiguration, 'publicIPAddressResourceId') ? {
       id: ipConfiguration.publicIPAddressResourceId
     } : null
-    subnet: contains(ipConfiguration, 'subnetResourceId') ? {
-      id: ipConfiguration.subnetResourceId
-    } : null
   }
 }]
 
-@description('Optional. The name of the diagnostic setting, if deployed.')
-param diagnosticSettingsName string = '${name}-diagnosticSettings'
+// ----------------------------------------------------------------------------
+// The reason we need to create a separate ipConfigurations object for each
+// scenario is because you cannot pass in
+//    publicIPAddress: {
+//      id: null
+//    }
+// because it will complain with the error: 'Value for reference ID is missing. Path
+// properties.ipConfigurations[0].properties.publicIpAddress
+// Otherwise, we could simply do
+//    publicIPAddress: {
+//      id: !(empty(azureFirewallSubnetPublicIpId)) ? azureFirewallSubnetPublicIpId : (isCreateDefaultPublicIP ? publicIPAddress.outputs.resourceId : null)
```

Comment:
We should be able to make this work using the `union` function. Let's have a chat some time

Reply:
keeping note of a commit where this is working 66ffbc2c2f0d7da8301366d664b72d51fb48184e - going to try the union logic now

Reply:
looks like union works. Please let me know if this is what you had in mind. Let's make it crisp so it'll be consistent on all other modules